StonehearthEditor/Recipes/RecipesView.cs

```diff
@@ -0,0 +1,701 @@
+            JObject json = jsonFileData.Json;
+
+            JArray newIngrArray = new JArray();
+            foreach (Ingredient ingredient in row.Ingredients)
+            {
+                if (!string.IsNullOrEmpty(ingredient.Name))
+                {
+                    string ingrKey = ingredient.Name.Contains(':') ? "uri" : "material";
+                    JObject newIngr = new JObject(
+                        new JProperty(ingrKey, ingredient.Name),
+                        new JProperty("count", ingredient.Amount));
+
+                    newIngrArray.Add(newIngr);
+                }
+            }
+
+            json["ingredients"] = newIngrArray;
+            jsonFileData.TrySetFlatFileData(json.ToString());
+            jsonFileData.TrySaveFile();
+        }
+
+        // Load the material constants from the stonehearth mod. This needs to be done before the data columns
+        // are populated,since they use this map to fill in data grid for material ingredients
+        private void LoadMaterialImages()
+        {
+            Module sh = ModuleDataManager.GetInstance().GetMod("stonehearth");
+            ModuleFile resourceConstants = sh.GetAliasFile("data:resource_constants");
+            if (resourceConstants != null)
+            {
+                JsonFileData jsonFileData = resourceConstants.FileData as JsonFileData;
+                JObject resourceConstantsJson = jsonFileData.Json;
+                JObject resourceJson = resourceConstantsJson["resources"] as JObject;
+
+                foreach (JProperty resource in resourceJson.Properties())
+                {
+                    string name = resource.Name;
+                    string iconLocation = resource.Value["icon"].ToString();
+                    string path = JsonHelper.GetFileFromFileJson(iconLocation, jsonFileData.Directory);
+                    if (path != "" && System.IO.File.Exists(path))
+                    {
+                        mMaterialImages.Add(name, path);
+                    }
+                }
+            }
+        }
+
+        private void SetGridValue(int colIndex, int rowIndex, string value)
+        {
+            try
+            {
+                if (recipesGridView.Columns[colIndex].ValueType == typeof(int))
+                {
+                    recipesGridView[colIndex, rowIndex].Value = int.Parse(value);
+                }
+                else
+                {
+                    bool valid = true;
+                    if (recipesGridView.Columns[colIndex] is DataGridViewComboBoxColumn)
+                    {
+                        List<string> autoCompleteStrings = (recipesGridView.Columns[colIndex] as DataGridViewComboBoxColumn).DataSource as List<string>;
+                        if (!autoCompleteStrings.Contains(value))
+                            valid = false;
+                    }
+
+                    if (valid)
+                        recipesGridView[colIndex, rowIndex].Value = value;
```

Comment:
If I paste an ingredient name, the icon doesn't update to match.

Reply:
the icon doesn't change until after you leave the cell/control and then columnchanged event gets triggered after that